Fix/dpp wpa msg fail

fengming-ye opened the pull request:
modules: hostap: remove zephyr wpa_msg socket pair
    
    Background:
    In DPP hostapd will put more than 1K bytes text in multiple
    wpa_msg calls in one loop. This will drain hostapd monitor
    socket pair and fail. Thus DPP connection will fail.
    
    The monitor socket pair of wpa_supplicant and hostapd are not like
    ctrl_iface socket pair, which are used for sending command to
    hostap from other tasks. The monitor socket pair is used by wpa_msg
    to send messages to attatched ctrl_iface monitors. But in zephyr case,
    the monitor is hostap task itself.
    This means hostap task is both the sender and receiver task.
    So it writes all the wpa_msg text into FIFO and read it in next loop.
    So there is risk of draining socket pair FIFO and missing msg.
    And the reading socket loop can be omitted by directly handling msg
    in wpa_msg_cb.
    
    Linux uses monitor socket pair because hostap is a process and wpa_msg
    does cross-process communication.
    But zephyr hostap is used as module in the same binary with
    zephyr kernel.
    So the usage is different and we don't need to use socket pair to
    notify ctrl_iface monitors.
    As long as we don't do time-consuming process in zephyr_wpa_msg_cb,
    it won't affect hostap task.
    So the zephyr_wpa_msg_cb will only filter and restore interesting logs,
    or raise mgmt events to wifi l2 mgmt task.
    
    Fix:
    Remove socket send flow in wpa_msg.
    Directly filter and handle text in zephyr_wpa_msg_cb, coding in zephyr
    repo.
    This will help us save 8K RAM of socket pair FIFO
    and max 50% loops of hostap task.

fengming-ye committed:
modules: hostap: remove zephyr wpa_msg socket pair

Background:
In DPP hostapd will put more than 1K bytes text in multiple
wpa_msg calls in one loop. This will drain hostapd monitor
socket pair and fail. Thus DPP connection will fail.

The monitor socket pair of wpa_supplicant and hostapd are not like
ctrl_iface socket pair, which are used for sending command to
hostap from other tasks. The monitor socket pair is used by wpa_msg
to send messages to attatched ctrl_iface monitors. But in zephyr case,
the monitor is hostap task itself.
This means hostap task is both the sender and receiver task.
So it writes all the wpa_msg text into FIFO and read it in next loop.
So there is risk of draining socket pair FIFO and missing msg.
And the reading socket loop can be omitted by directly handling msg
in wpa_msg_cb.

Linux uses monitor socket pair because hostap is a process and wpa_msg
does cross-process communication.
But zephyr hostap is used as module in the same binary with
zephyr kernel.
So the usage is different and we don't need to use socket pair to
notify ctrl_iface monitors.
As long as we don't do time-consuming process in zephyr_wpa_msg_cb,
it won't affect hostap task.
So the zephyr_wpa_msg_cb will only filter and restore interesting logs,
or raise mgmt events to wifi l2 mgmt task.

Fix:
Remove socket send flow in wpa_msg.
Directly filter and handle text in zephyr_wpa_msg_cb,
coding in zephyr repo.
This will save 2K RAM in supplicant case and 4K RAM in
supplicant & hostapd coex case, in current default hostap
socket pair config.
And this will save max 50% loops of hostap task.

Signed-off-by: Fengming Ye <frank.ye@nxp.com>

fengming-ye committed:
manifest: update hostap revision

Update hostap revision to fix wifi dpp fail due to
wpa_msg draining socket pair FIFO.

Signed-off-by: Fengming Ye <frank.ye@nxp.com>

krish2718 commented:
> This means hostap task is both the sender and receiver task. 

Well, wpa_msg is littered throughout the code and can be called from multiple threads, I guess when we use control interface + eloop we have those multiple messages are serialized, but I still wouldn't call the current design thread safe, so, removing sockets should be fine, but I need to do a deeper reivew  and also run some tests, will update soon.

fengming-ye commented:
> Well, wpa_msg is littered throughout the code and can be called from multiple threads

In my code view, I find wpa_msg is only called inside supplicant native code.
And native code is all kept in only eloop thread, to be thread-safe.
This is the purpose of zephyr hostap initial arch design from your side, I guess.

Could you share some example of wpa_msg called from other thread?

fengming-ye commented:
modified to improve that in either supplicant or supplicant & hostapd case, we should register zephyr_hostap_ctrl_iface_msg_cb and zephyr_hostap ifname_cb

krish2718 commented:
> > Well, wpa_msg is littered throughout the code and can be called from multiple threads
> 
> In my code view, I find wpa_msg is only called inside supplicant native code. And native code is all kept in only eloop thread, to be thread-safe. This is the purpose of zephyr hostap initial arch design from your side, I guess.

Yes, sockets + eloop is the way to go to implement anything in supplicant to be thread safe.
> 
> Could you share some example of wpa_msg called from other thread?

I see some usage in `ctrl_iface.c` but not common events, so, may be we can ignore those for now?

> As long as we don't do time-consuming process in zephyr_wpa_msg_cb,
it won't affect hostap task.

I guess this is fine as all we are doing is sending a net_mgmt event which just queues and returns (the actual sending is done in net_mgmt thread).

BTW, our local tests passed on this, so, I am approving

fengming-ye commented:
> @fengming-ye GH says there is a conflict, could you check it?

It's because west.yml hostap revision updated.
And I want to hold this PR until Hui's PR https://github.com/zephyrproject-rtos/zephyr/pull/83253 is merged.
There will be a code location change based on that PR. And I don't want to conflict that PR because that has been hold too long.

jukkar commented:
> There will be a code location change based on that PR. And I don't want to conflict that PR because that has been hold too long.

Ok, no worries, that is fine.

krish2718 commented:
Should we fine-tune https://github.com/zephyrproject-rtos/zephyr/blob/42c3a7814810a60cd23d719264695351ba5b8023/subsys/net/lib/sockets/Kconfig#L354 `NET_SOCKETPAIR_MAX` and `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` ?

krish2718 commented:
> Should we fine-tune
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/42c3a7814810a60cd23d719264695351ba5b8023/subsys/net/lib/sockets/Kconfig#L354
> 
> `NET_SOCKETPAIR_MAX` and `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` ?

I have tried to build with and without this PR, but I don't see any decrease in RAM, can you please share some details?

fengming-ye commented:
> > Should we fine-tune
> > https://github.com/zephyrproject-rtos/zephyr/blob/42c3a7814810a60cd23d719264695351ba5b8023/subsys/net/lib/sockets/Kconfig#L354
> > 
> > `NET_SOCKETPAIR_MAX` and `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` ?
> 
> I have tried to build with and without this PR, but I don't see any decrease in RAM, can you please share some details?

`NET_SOCKETPAIR_MAX` is for static socketpair alloc.
We are using heap malloc and I can tune `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` to reduce 2K for supplicant and 4K for supplicant & hostapd coex cases.

krish2718 commented:
> > > Should we fine-tune
> > > https://github.com/zephyrproject-rtos/zephyr/blob/42c3a7814810a60cd23d719264695351ba5b8023/subsys/net/lib/sockets/Kconfig#L354
> > > 
> > > `NET_SOCKETPAIR_MAX` and `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` ?
> > 
> > 
> > I have tried to build with and without this PR, but I don't see any decrease in RAM, can you please share some details?
> 
> `NET_SOCKETPAIR_MAX` is for static socketpair alloc. We are using heap malloc and I can tune `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` to reduce 2K for supplicant and 4K for supplicant & hostapd coex cases.

so, it's a 2K RAM improvement for supplicant only?

fengming-ye commented:
> > > > Should we fine-tune
> > > > https://github.com/zephyrproject-rtos/zephyr/blob/42c3a7814810a60cd23d719264695351ba5b8023/subsys/net/lib/sockets/Kconfig#L354
> > > > 
> > > > `NET_SOCKETPAIR_MAX` and `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` ?
> > > 
> > > 
> > > I have tried to build with and without this PR, but I don't see any decrease in RAM, can you please share some details?
> > 
> > 
> > `NET_SOCKETPAIR_MAX` is for static socketpair alloc. We are using heap malloc and I can tune `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` to reduce 2K for supplicant and 4K for supplicant & hostapd coex cases.
> 
> so, it's a 2K RAM improvement for supplicant only?

yes

krish2718 commented:
> > > > > Should we fine-tune
> > > > > https://github.com/zephyrproject-rtos/zephyr/blob/42c3a7814810a60cd23d719264695351ba5b8023/subsys/net/lib/sockets/Kconfig#L354
> > > > > 
> > > > > `NET_SOCKETPAIR_MAX` and `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` ?
> > > > 
> > > > 
> > > > I have tried to build with and without this PR, but I don't see any decrease in RAM, can you please share some details?
> > > 
> > > 
> > > `NET_SOCKETPAIR_MAX` is for static socketpair alloc. We are using heap malloc and I can tune `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` to reduce 2K for supplicant and 4K for supplicant & hostapd coex cases.
> > 
> > 
> > so, it's a 2K RAM improvement for supplicant only?
> 
> yes

Can you please update commits logs to reflect the same? Else it will cause confusion.

fengming-ye commented:
> > > > > > Should we fine-tune
> > > > > > https://github.com/zephyrproject-rtos/zephyr/blob/42c3a7814810a60cd23d719264695351ba5b8023/subsys/net/lib/sockets/Kconfig#L354
> > > > > > 
> > > > > > `NET_SOCKETPAIR_MAX` and `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` ?
> > > > > 
> > > > > 
> > > > > I have tried to build with and without this PR, but I don't see any decrease in RAM, can you please share some details?
> > > > 
> > > > 
> > > > `NET_SOCKETPAIR_MAX` is for static socketpair alloc. We are using heap malloc and I can tune `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` to reduce 2K for supplicant and 4K for supplicant & hostapd coex cases.
> > > 
> > > 
> > > so, it's a 2K RAM improvement for supplicant only?
> > 
> > 
> > yes
> 
> Can you please update commits logs to reflect the same? Else it will cause confusion.

okay I will update commits after hostapd split PR merged.

MaochenWang1 commented:
please also change the https://github.com/zephyrproject-rtos/zephyr/blob/00cc4441aa2313f2cca721b94db9bce86b0af22c/subsys/net/lib/sockets/Kconfig#L363
as the number of socketpair is reduced

fengming-ye committed:
net: sockets: kconfig: change hostap default socket pair add heap size

Reduce default socket pair add heap size in hostap case.
Reduce 1 socket pair in wpa_supplicant case.
Reduce 2 socket pairs in wpa_supplicant and hostapd coex case.

Signed-off-by: Fengming Ye <frank.ye@nxp.com>

fengming-ye commented:
rebased on hostapd split change and reduce socketpair add heap size kconfig option for hostap

jukkar commented on modules/hostap/src/supp_main.c:
```diff
@@ -227,6 +227,80 @@ static int get_iface_count(struct supplicant_context *ctx)
+	}
+
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+	if ((*((int *)ctx)) == 0) {
```
This is technically correct but I think we need to change this to more user friendly way as we are now moving/changing these lines. It would be better (and self documenting) if we do it like this instead
```
if ((((struct wpa_supplicant *)ctx))->is_hostapd == 0) {
```
This allows user to understand what is going on in the checks. I just spent few minutes trying to understand what we are doing with the checks.

fengming-ye replied:
ok updated

jukkar commented on modules/hostap/src/supp_main.c:
```diff
@@ -227,6 +227,80 @@ static int get_iface_count(struct supplicant_context *ctx)
+	}
+
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+	if ((*((int *)ctx)) == 0) {
```
same here

fengming-ye replied:
ok updated

jukkar commented:
I just merged the hostap PR, you can update the manifest.

krish2718 commented:
> > > > > > Should we fine-tune
> > > > > > https://github.com/zephyrproject-rtos/zephyr/blob/42c3a7814810a60cd23d719264695351ba5b8023/subsys/net/lib/sockets/Kconfig#L354
> > > > > > 
> > > > > > `NET_SOCKETPAIR_MAX` and `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` ?
> > > > > 
> > > > > 
> > > > > I have tried to build with and without this PR, but I don't see any decrease in RAM, can you please share some details?
> > > > 
> > > > 
> > > > `NET_SOCKETPAIR_MAX` is for static socketpair alloc. We are using heap malloc and I can tune `HEAP_MEM_POOL_ADD_SIZE_SOCKETPAIR` to reduce 2K for supplicant and 4K for supplicant & hostapd coex cases.
> > > 
> > > 
> > > so, it's a 2K RAM improvement for supplicant only?
> > 
> > 
> > yes
> 
> Can you please update commits logs to reflect the same? Else it will cause confusion.

Please update
